artiq/coredevice/urukul.py

```diff
@@ -124,6 +135,492 @@ def set_mu(self, ftw: TInt32):
+        self._not_implemented()
+
+    @kernel
+    def cfg_drctl(self, cpld, channel: TInt32, on: TBool):
+        self._not_implemented()
+
+    @kernel
+    def cfg_drctl_all(self, cpld, state: TInt32):
+        self._not_implemented()
+
+    @kernel
+    def cfg_drhold(self, cpld, channel: TInt32, on: TBool):
+        self._not_implemented()
+
+    @kernel
+    def cfg_drhold_all(self, cpld, state: TInt32):
+        self._not_implemented()
+
+
+class ProtoRev8(CPLDVersion):
+    """
+    Implementation of the CPLD for Urkul ProtoRev8.
+    """
+
+    # ProtoRev8 CFG configuration register bit offsets
+    CFG_IO_UPDATE = 12
+    CFG_MASK_NU = 13
+    CFG_CLK_SEL0 = 17
+    CFG_CLK_SEL1 = 21
+    CFG_SYNC_SEL = 18
+    CFG_RST = 19
+    CFG_IO_RST = 20
+    CFG_CLK_DIV = 22
+
+    @staticmethod
+    @portable
+    def urukul_cfg(
+        rf_sw,
+        led,
+        profile,
+        io_update,
+        mask_nu,
+        clk_sel,
+        sync_sel,
+        rst,
+        io_rst,
+        clk_div,
+    ):
+        """Build Urukul CPLD configuration register"""
+        return (
+            (rf_sw << CFG_RF_SW)
+            | (led << CFG_LED)
+            | (profile << CFG_PROFILE)
+            | (io_update << ProtoRev8.CFG_IO_UPDATE)
+            | (mask_nu << ProtoRev8.CFG_MASK_NU)
+            | ((clk_sel & 0x01) << ProtoRev8.CFG_CLK_SEL0)
+            | ((clk_sel & 0x02) << (ProtoRev8.CFG_CLK_SEL1 - 1))
+            | (sync_sel << ProtoRev8.CFG_SYNC_SEL)
+            | (rst << ProtoRev8.CFG_RST)
+            | (io_rst << ProtoRev8.CFG_IO_RST)
+            | (clk_div << ProtoRev8.CFG_CLK_DIV)
+        )
+
+    @kernel
+    def cfg_write(self, cpld, cfg: TInt32):
+        """Write to the configuration register.
+
+        See :func:`urukul_cfg` for possible flags.
+
+        :param cfg: 24-bit data to be written. Will be stored at
+            :attr:`cfg_reg`.
+        """
+        cpld.bus.set_config_mu(SPI_CONFIG | spi.SPI_END, 24, SPIT_CFG_WR, CS_CFG)
+        cpld.bus.write(cfg << 8)
+        cpld.cfg_reg = cfg
+
+    @kernel
+    def sta_read(self, cpld) -> TInt32:
+        """Read the status register.
+
+        Use any of the following functions to extract values:
+
+            * :func:`urukul_sta_rf_sw`
+            * :func:`urukul_sta_smp_err`
+            * :func:`urukul_sta_pll_lock`
+            * :func:`urukul_sta_ifc_mode`
+            * :func:`urukul_sta_proto_rev`
+
+        :return: The status register value.
+        """
+        cpld.bus.set_config_mu(
+            SPI_CONFIG | spi.SPI_END | spi.SPI_INPUT, 24, SPIT_CFG_RD, CS_CFG
+        )
+        cpld.bus.write(cpld.cfg_reg << 8)
+        return cpld.bus.read()
+
+    @kernel
+    def init(self, cpld):
+        """Initialize Urukul with ProtoRev8.
+
+        Resets the DDS I/O interface.
+        Does not pulse the DDS ``MASTER_RESET`` as that confuses the AD9910.
+        """
+        if cpld.proto_rev != STA_PROTO_REV_8:
```

Comment:
I meant, can you read STA from the Urukul itself to compare the proto rev? This test here will always pass as you set ``proto_rev`` in __init__ yourself anyway.

Reply:
Well that is slightly embarrassing.  When I had read your original message I knew what you were talking about and in my mind I was planning on doing that but when I got around to implementing it I went blank on that.  Fix pushed.